### PR TITLE
[Tensorflow] Several nn & example 

### PR DESCRIPTION
**Pull request by VoVAllen**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

File: examples/tensorflow/gat/utils.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import numpy as np
+import torch
```
Comment (yzh119) [inline examples/tensorflow/gat/utils.py]: Remove this.

File: tests/tensorflow/test_nn.py [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+    assert F.allclose(a, uniform_attention(g, a.shape))
+
+    # Test higher dimension case
+    edata = F.ones((g.number_of_edges(), 3, 1))
+    a = nn.edge_softmax(g, edata)
+    assert len(g.ndata) == 0
+    assert len(g.edata) == 0
+    assert F.allclose(a, uniform_attention(g, a.shape))
+
+    # Test both forward and backward with PyTorch built-in softmax.
```
Comment (yzh119) [inline tests/tensorflow/test_nn.py]: Try using sed to replace all the occurance of torch with tf.

File: examples/tensorflow/rgcn/utils.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    values = np.random.randint(num_entity, size=num_to_generate)
+    choices = np.random.uniform(size=num_to_generate)
+    subj = choices > 0.5
+    obj = choices <= 0.5
+    neg_samples[subj, 0] = values[subj]
+    neg_samples[obj, 2] = values[obj]
+
+    return np.concatenate((pos_samples, neg_samples)), labels
+
+#######################################################################
```
Comment (yzh119) [inline examples/tensorflow/rgcn/utils.py]: Why disable following functions?

Comment (VoVAllen) [reply examples/tensorflow/rgcn/utils.py]: Because I only implemented entity classification. Link prediction is not implemented yet.

**Comment by yzh119** [comment]
Please remove the unused codes and then I'll approve this PR.

File: python/dgl/nn/tensorflow/conv/relgraphconv.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
+
```
Comment (yzh119) [inline python/dgl/nn/tensorflow/conv/relgraphconv.py]: Remove this line.

File: examples/tensorflow/gat/README.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+
+Results
+-------
+
+| Dataset  | Test Accuracy |
+| -------- | ------------- |
+| Cora     | 84.2          |
+| Citeseer | 70.9          |
+| Pubmed   | 78.5          |
```
Comment (jermainewang) [inline examples/tensorflow/gat/README.md]: Are there any baseline numbers?

**Comment by jermainewang** [comment]
I didn't see a README for RGCN btw.

File: examples/tensorflow/gat/README.md (hunk 5 above)
Comment (VoVAllen) [reply examples/tensorflow/gat/README.md]: Added

File: python/dgl/nn/tensorflow/conv/relgraphconv.py (hunk 4 above)
Comment (VoVAllen) [reply python/dgl/nn/tensorflow/conv/relgraphconv.py]: Done

**Comment by VoVAllen** [comment]
@jermainewang I added the rgcn README. TF works well on AIFB and MUTAG. But for BGS it goes OOM  when n-base=30. Decreasing n-base to 25 works on GPU works well but results in worse result. However, when using n-base=30 on CPU, it still gives out worse result comparing to PyTorch. I'm still investigating the reason.

File: examples/tensorflow/rgcn/README.md [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+```
+python3 entity_classify.py -d aifb --testing --gpu 0
+```
+
+MUTAG: accuracy 75% (DGL), 73.23% (paper)
+```
+python3 entity_classify.py -d mutag --l2norm 5e-4 --n-bases 30 --testing --gpu 0
+```
+
+BGS: accuracy 79.3% (DGL n-base=25), 83.10% (paper n-base=40)
```
Comment (jermainewang) [inline examples/tensorflow/rgcn/README.md]: You said DGL/TF run OOM for BGS?

**Comment by jermainewang** [comment]
Thanks for the README update. Is there any speed comparison that could be included in the README?

**Comment by VoVAllen** [comment]
> Thanks for the README update. Is there any speed comparison that could be included in the README?

I found speed varies up to 40% when I use different version of tensorflow(tf-nightly-gpu vs tensorflow 2.1.0). And I hadn't carefully tested the speed for now.

**Comment by sjs2109** [comment]
hello @VoVAllen 
I've tested GAT tensorflow version in google colab but it is not working
error code is as below

Downloading /root/.dgl/citeseer.zip from https://data.dgl.ai/dataset/citeseer.zip...
Extracting file to /root/.dgl/citeseer
/usr/local/lib/python3.6/dist-packages/dgl/data/citation_graph.py:140: RuntimeWarning: divide by zero encountered in power
  r_inv = np.power(rowsum, -1).flatten()
Finished data loading and preprocessing.
  NumNodes: 3327
  NumEdges: 9228
  NumFeats: 3703
  NumClasses: 6
  NumTrainingSamples: 120
  NumValidationSamples: 500
  NumTestSamples: 1000
> <ipython-input-3-a16ba8030e58>(42)main()
-> if args.gpu < 0:
(Pdb) c
----Data statistics------'
        #Edges 9228
        #Classes 6 
        #Train samples 120
        #Val samples 500
        #Test samples 1000
<__main__.GAT object at 0x7f22c51af860>
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-7-b3f47c9a9b7f> in <module>()
----> 1 main(args)

9 frames
<ipython-input-3-a16ba8030e58> in main(args)
     40 
     41     pdb.set_trace()
---> 42     if args.gpu < 0:
     43         device = "/cpu:0"
     44     else:

/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/engine/base_layer.py in __call__(self, *args, **kwargs)
    862           with base_layer_utils.autocast_context_manager(
    863               self._compute_dtype):
--> 864             outputs = self.call(cast_inputs, *args, **kwargs)
    865           self._handle_activity_regularization(inputs, outputs)
    866           self._set_mask_metadata(inputs, outputs, input_masks)

<ipython-input-2-af4cd9e7a474> in call(self, inputs)
     50         h = inputs
     51         for l in range(self.num_layers):
---> 52             h = self.gat_layers[l](self.g, h)
     53             h = tf.reshape(h, (h.shape[0], -1))
     54         # output projection

/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/engine/base_layer.py in __call__(self, *args, **kwargs)
    862           with base_layer_utils.autocast_context_manager(
    863               self._compute_dtype):
--> 864             outputs = self.call(cast_inputs, *args, **kwargs)
    865           self._handle_activity_regularization(inputs, outputs)
    866           self._set_mask_metadata(inputs, outputs, input_masks)

/usr/local/lib/python3.6/dist-packages/dgl/nn/tensorflow/conv/gatconv.py in call(self, graph, feat)
    108         graph.ndata.update({'ft': feat, 'el': el, 'er': er})
    109         # compute edge attention
--> 110         graph.apply_edges(fn.u_add_v('el', 'er', 'e'))
    111         e = self.leaky_relu(graph.edata.pop('e'))
    112         # compute softmax

/usr/local/lib/python3.6/dist-packages/dgl/graph.py in apply_edges(self, func, edges, inplace)
   2146         with ir.prog() as prog:
   2147             scheduler.schedule_apply_edges(AdaptedDGLGraph(self), u, v, eid, func, inplace)
-> 2148             Runtime.run(prog)
   2149 
   2150     def group_apply_edges(self, group_by, func, edges=ALL, inplace=False):

/usr/local/lib/python3.6/dist-packages/dgl/runtime/runtime.py in run(prog)
      9         for exe in prog.execs:
     10             # prog.pprint_exe(exe)
---> 11             exe.run()

/usr/local/lib/python3.6/dist-packages/dgl/runtime/ir/executor.py in run(self)
   1063         lhs_data = self.lhs_data.data
   1064         rhs_data = self.rhs_data.data
-> 1065         ctx = utils.to_dgl_context(F.context(lhs_data))
   1066         graph = self.graph.data(ctx)
   1067         lhs_map = self.lhs_map.data(ctx) if self.lhs_map.data else None

/usr/local/lib/python3.6/dist-packages/dgl/utils.py in to_dgl_context(ctx)
    496 def to_dgl_context(ctx):
    497     """Convert a backend context to DGLContext"""
--> 498     device_type = nd.DGLContext.STR2MASK[F.device_type(ctx)]
    499     device_id = F.device_id(ctx)
    500     return nd.DGLContext(device_type, device_id)

/usr/local/lib/python3.6/dist-packages/dgl/backend/pytorch/tensor.py in device_type(ctx)
     66 
     67 def device_type(ctx):
---> 68     return ctx.type
     69 
     70 def device_id(ctx):

AttributeError: 'str' object has no attribute 'type'

Could you kindly help to fix the issue?
Thanks

**Comment by VoVAllen** [comment]
@sjs2109 Hi, you need to set environment variable DGLBACKEND=tensorflow before importing dgl.
```python
import os
os.environ['DGLBACKEND'] = 'tensorflow'
# Then import dgl
import dgl
```

**Comment by VoVAllen** [comment]
@sjs2109 Please refer to this document https://github.com/dmlc/dgl/blob/f1a8f926a4ac3f9ffbc38fa3c57ca5c67a5cd652/docs/source/install/backend.rst. Also you need extra package tfdlpack and set some environment variable for Tensorflow usage.

**Comment by sjs2109** [comment]
Hi @VoVAllen 
Thanks for your kindly help.

**Pull request merged**